### ISSUE TITLE
Experiments planning

### ISSUE THREAD

**Issue by m09** [src-d/tm-experiments#1]
Research the most insightful ways to apply topic modeling to source code.

- [x] Review existing literature: https://www.overleaf.com/7163137971dqhdpqmkdhyv
- [x] Create the feature extraction and model training pipeline:
  - [x] Parse repositories at several points in history and extract features for each revision, following either the Diff or Hall model defined by Thomas (cf the literature review)
  - [x] Define a pipeline to compute a topic model over the complete features
  - [x] Post-process the obtained model to obtain topics for each file in each considered point in history
  - [ ] Explore the interaction of the diff model with document sparsity regularization and general topic model training by changing the corpus we train on: instead of keeping diff documents (both added and deleted) to train on, try to concatenate all add-documents of a given document into a consolidated document for training, and use the diff-documents only to compute topics for each revision in post-processing
- [ ] Make sure the models trained with the Python ARTM client are equivalent to the ones trained with the ARTM CLI.
- [ ] Experiment on https://github.com/pytorch/pytorch/ (20+ releases):
  - [ ] Find the best way to automatically label topics:
    - [x] Implement Mei's approach
    - [ ] Test different ways to define contexts for labeling: https://github.com/src-d/tm-experiments/pull/31#issuecomment-519007096, with the added option of using the max/mean/median count across revisions for each identifier as a context
    - [ ] Test monoword labels to mitigate the bigram mining/ordering problem
    - [ ] Test filtering label names to use only some type of identifiers (class names, function names, …), potentially boosting them compared to others when we score them using the trained topic model
  - [x] Research how to circumvent the need for a manually set number of topics
  - [ ] Measure topic evolution cohesion with existing changelogs
  - [x] Define the best evaluation schemes for production setups (which metrics, at which stages)
- [ ] Scale up the experiments on http://github.com/tensorflow/tensorflow/ (80+ releases), and to commit granularity. Currently blocked by:
  - bblfsh python client OOMs
  - gitbase UAST cache issues/OOMs when parsing UASTs
  - diffs are not parsable in gitbase
- [ ] Compute the topics at the user level, not document level, using either blame to determine ownership (easy to try, not so precise) or identifiers in diffs (a bit harder, very precise), and apply all the above points to the obtained topics (automatic labels, evolution, …)

**Comment by m09** [comment]
@r0mainK I've added a few points on feature extraction/model building, feel free to edit/enhance/…!

**Comment by vmarkovtsev** [comment]
Please note: we should use BigARTM for everything around topic modeling for multiple reasons:
- Much more handles and parameters to explore
- We are in good control of the code, the maintainers are quite nice and we do not have issues with fixing bugs
- Related to (2), when something does not work, the maintainers are always helpful
- The code is better than gensim

**Comment by m09** [comment]
This is a research exploration, I don't think we should pinpoint the exact software we use, it might limit the amount of insight we get into the problem. Also, there are 2 big problems related to BigARTM for this experiment:
- there is no non-parametric method implemented
- the code is mainly documented in Russian, which is a problem for more than half of the team

But indeed it's a good and efficient project and we already planned to use it it was at suitable.

**Comment by vmarkovtsev** [comment]
As Romain demonstrated, the non-parametric method failed miserably, so this is no longer a problem :smile: 

I am mostly thinking about future use in production and trying to reduce the number of artificial rewrites. From what I saw in the leading R&D labs in the past, it is much beneficial to enforce such software constraints earlier than later. The alternatives should be comparable, however.

If we rely on a unique property of some specific library, and we require another one in production due to some research-irrelevant reasons, this may be a huge problem on a later stage. The difference between the research team at source{d} and a typical university research group is that our sole research purpose is to help the downstream folks commercialize.

Returning back from the metaphysical discussions: we should use a non-production-friendly library for a really, really serious reason. E.g. we are exercising PyTorch because GGNNs are an infinite pain to **experiment with** in Tensorflow.

**Comment by m09** [comment]
As I mentioned in the previous comment we did plan to use BigARTM if it was feasible. But let's take the result that @r0mainK obtained: it was not obtainable quickly with BigARTM.

I agree with the conclusion of your second message :+1: (with the note that there are often serious reasons to depart from a lib in experiments).

**Comment by m09** [comment]
from #18 by @r0mainK:
> I actually rethought about it, I think it would probably make more sense to do the following:
>
> 1. take a much lower sparsity tolerance for doctopic, like 1%, which will reflect better reality,
> 2. distinguish that tolerance from the prob used to get non-superfluous topic, which should be higher
>
> I think we should simply throw topics which are not at say 30 or even 50% dominance in at least one file. Also, it doesn't necessarily have to be one file, depending on the corpus, so we should add that as a parameter. By the way, using this config (over 50% in at least one file), I get only 17 topics with the perplexity metric. Going to do the changes in a new commit, and test for both metrics, tell me what you think of this idea ?

**Comment by r0mainK** [comment]
So in the end we talked about this on Slack, here is the jist:

- agreed that 10% is a bit high when printing sparsity, settled on 5%
- agreed on separating value used to compute score and value use to remove/select topics
- given that the current strategy amount to remove topics that don't have at least probability of 0.1 in at least one document and apparent success of experiments, [next PR](https://github.com/src-d/tm-experiments/pull/19) will generalize this with harder threshold (0.5) and add arguments regarding min number of documents (either absolute or relative to the number of documents/delta-documents)
- in the future we will add selection based on entropy, which requires some hyperparameter opt - probably a threshold which should be chosen from the entropies distribution after using the selectin reg

**Comment by m09** [comment]
@r0mainK I updated the first comment according to https://github.com/src-d/minutes/pull/685/files, feel free to add/modify/delete any point :upside_down_face:

**Comment by r0mainK** [comment]
**Update**

Just a small update on the progress:
- see issue #33 , I identified some more bugs (from the code, but also BigARTM), and implemented fixes/workarounds. Unfortunately, this meant removing perplexity score. As it is not completely straightforward to implement, I will not do it right now (especially since `distinctness` seems to be a better metric)
- I added the consolidated corpus logic in `create-bow` and `train-$`, we can now start experimenting on this
- I implemented the different context creation methods for labeling, we can now start experimenting on this
- I started discussing with LA the need for `diffs`, it seems there already exists an implementation (that is not exposed) however it is not really suited for our needs. I proposed an alternative way of doing it which I expect is unfeasible given what I learnt from Babelfish in that conversation, however hopefully the API  design proposed will not.

**Next steps**

- [x] Finish multi-repo refactoring:
    - [x] postprocess: 2 outputs currently, the doc membership (immediate for HALL, not so much for DIFF) and total word count, which could be move to create-bow (not much overhead, but duplicate), add typing
    - [x] metrics: add multi-repo, maybe typing
    - [x] visualize: add multi-repo, maybe typing
- [ ] Once the new babelfish python-client is out, preprocess `pytorch` once more and start experimenting seriously - and also hopefully remove some of the restart logic.
- [ ] Once the new Gitbase is out (with hard limit on cache), experiment to see if it can be used to make preprocessing way faster, possibly doing most of the work in it through SQL queries (less I/O client side)
 - [ ] A PR in Gitbase is currently under way to expose `blame` function. Once it is merged, we can start implementing user ownership evolution.
